platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java

```diff
@@ -0,0 +1,210 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(SolrPasswordUpdateImpl.class);
+
+  @SuppressWarnings("squid:S2068" /* This constant does hold an actual password */)
+  private static final String SOLR_PASSWORD_PROPERTY_NAME = "solr.password";
+
+  private final ClientFactoryFactory restClientFactoryFactory;
+  private final UuidGenerator uuidGenerator;
+  private java.util.Properties properties;
+  private SolrAuthResource solrAuthResource;
+  private String newPasswordPlainText;
+  private String newPasswordEncrypted;
+  private EncryptionService encryptionService;
+  private StatusType solrResponse;
+  private String newPasswordWrappedEncrypted;
+
+  public SolrPasswordUpdateImpl(
+      UuidGenerator uuidGenerator,
+      ClientFactoryFactory clientFactoryFactory,
+      EncryptionService encryptionService) {
+    this.encryptionService = encryptionService;
+    this.uuidGenerator = uuidGenerator;
+    this.restClientFactoryFactory = clientFactoryFactory;
+    properties =
+        AccessController.doPrivileged(
+            (PrivilegedAction<java.util.Properties>) System::getProperties);
+  }
+
+  private void initialize() {
+    SecureCxfClientFactory<SolrAuthResource> factory =
+        restClientFactoryFactory.getSecureCxfClientFactory(
+            properties.getProperty("solr.http.url"),
+            SolrAuthResource.class,
+            properties.getProperty("solr.username"),
+            getPlaintextPasswordFromProperties());
+    solrAuthResource = factory.getClient();
+  }
+
+  /**
+   * If the password is not changed, change it. Only call this method if the system is configured to
+   * use basic authentication for Solr. This method will only change the password if it detects that
+   * the default password is still in use. This object is intended to be a singleton, a blueprint
+   * bean. That is why is can use itself as the lock
+   */
+  public synchronized void updateSolrPassword() {
+    if (configuredToAttemptAutoPasswordChange() && isUsingDefaultPassword()) {
+      initialize();
+      generatePassword();
+      setPasswordInSolr();
+      if (isSolrPasswordChangeSuccessfull()) {
+        setPasswordInMemory();
+        setPasswordInPropertiesFile();
+      }
+    }
+  }
+
+  private void setPasswordInMemory() {
+    properties.setProperty(SOLR_PASSWORD_PROPERTY_NAME, newPasswordWrappedEncrypted);
+    LOGGER.info("Updated encrypted Solr password in memory.");
+  }
+
+  private boolean configuredToAttemptAutoPasswordChange() {
+    return Boolean.valueOf(properties.getProperty("solr.attemptAutoPasswordChange"));
+  }
+
+  /**
+   * The return value of this method will change after setPasswordInMemory() is called
+   *
+   * @return true if in-memory password is the same as hard coded default password
+   */
+  private boolean isUsingDefaultPassword() {
+    return getPlaintextPasswordFromProperties().equals("autogenerated");
```

Comment:
❗️ `getPlaintextPasswordFromProperties()` is marked as `@Nullable` which means this can result in a NPE. Reverse the check ... `"autogenerated".equals(getPlaintextPasswordFromProperties())`